upstairs/src/lib.rs

```diff
@@ -481,41 +477,49 @@ async fn proc(
                                 DsState::BadVersion
                             );
                             bail!("expected version 1, got {}", version);
+                        } else {
+                            up.ds_transition(
+                                up_coms.client_id,
+                                DsState::CompatVersion
+                            );
                         }
-
-                        negotiated = 1;
-
-                        // May be some time between promotion to active, so ping until then
-                        needping = true;
                     }
                     Some(Message::Imok) => {
-                        println!("{} client {} is waiting for promotion to active.", up.uuid, up_coms.client_id);
+                        if up.ds_state(up_coms.client_id) == DsState::CompatVersion {
+                            println!("{} client {} is waiting for promotion to active, received ping response.",
+                                up.uuid, up_coms.client_id
+                            );
+                        }
                     }
                     Some(Message::YouAreNowActive(uuid)) => {
                         assert_eq!(uuid, up.uuid);
-                        up.set_active();
+
+                        if up.ds_state(up_coms.client_id) != DsState::CompatVersion {
+                            bail!("received YouAreNowActive in state {:?}", up.ds_state(up_coms.client_id));
+                        }
+
+                        up.ds_transition(up_coms.client_id, DsState::Activated);
 
                         /*
                          * Get region info.
                          */
                         fw.send(Message::RegionInfoPlease).await?;
                     }
                     Some(Message::RegionInfo(region_def)) => {
-                        if negotiated != 1 {
-                            bail!("Received RegionInfo out of order!");
+                        if up.ds_state(up_coms.client_id) != DsState::Activated {
+                            bail!("received RegionInfo in state {:?}", up.ds_state(up_coms.client_id));
                         }
 
                         up.add_downstairs(up_coms.client_id, region_def)?;
-                        negotiated = 2;
 
                         /*
                          * Ask for the current version of all extents.
                          */
                         fw.send(Message::ExtentVersionsPlease).await?;
                     },
                     Some(Message::ExtentVersions(versions)) => {
-                        if negotiated != 2 {
-                            bail!("Received ExtentVersions out of order!");
+                        if up.ds_state(up_coms.client_id) != DsState::Activated {
```

Comment:
I think we want to require a downstairs to strictly follow the state transitions, which is what I was trying
with the hack `negotiated != 2` check.  
However, I think adding state transition work is too much more for this review, and can be done in a coming review.

Reply:
Did something in cd0d207, though not the formal state transition stuff we talked about

Reply:
Actually, I'm going to pull this out to another PR.